gosdkexample/generator/main.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/gosdk"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/gosdk/command"
+)
+
+// This example is a re-writing of https://github.com/GoogleContainerTools/kpt/issues/2435.
+// We generate with a strong typed object.
+
+func main() {
+	cmd := command.AsMain(context.TODO(), generate, command.Options{})
+	cmd.Short = "short description"
+	cmd.Long = "long description"
+	cmd.Example = "example"
+	if err := cmd.Execute(); err != nil {
+		os.Exit(1)
+	}
+}
+
+type Dashboard struct {
+	Revision string `json:"revision"`
+	ID       string `json:"id"`
+}
+
+func generate(ctx context.Context, rl *gosdk.ResourceList) error {
+	if rl.FunctionConfig == nil || rl.FunctionConfig.Node() == nil {
+		return fmt.Errorf("unable for find `functionConfig`")
+	}
+	var data map[string]string
+	found, err := rl.FunctionConfig.Get(&data, "data")
+	if err != nil {
+		return err
+	}
+	if !found {
+		return fmt.Errorf("unable for find `data` field in functionConfig")
+	}
+	db := Dashboard{
+		Revision: data["revision"],
+		ID:       data["id"],
+	}
+	js, err := fetchDashboard(db)
+	if err != nil {
+		return fmt.Errorf("fetch dashboard: %v", err)
+	}
+
+	cm := corev1.ConfigMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ConfigMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rl.FunctionConfig.GetName(),
+			Namespace: rl.FunctionConfig.GetNamespace(),
+			Labels: map[string]string{
+				"grafana_dashboard": "true",
+			},
+		},
+		Data: map[string]string{
+			fmt.Sprintf("%v.json", rl.FunctionConfig.GetName()): fmt.Sprintf("%q", js),
+		},
+	}
+	return rl.AddObjectToItems(cm)
```

Comment:
Interesting thing about generators is what happens if the object is already generated.  That's the thing that is hard to solve and maybe we should provide plumbing to do that.

Reply:
beyond the scope of this revision to the SDK though.

Reply:
I have updated `AddObjectToItems` and the examples